File: agent/tools/vercel.js
const axios = require('axios');

const VERCEL_API = 'https://api.vercel.com';

const headers = () => ({
  Authorization: `Bearer ${process.env.VERCEL_TOKEN}`,
  'Content-Type': 'application/json',
});

/**
 * Trigger a Vercel deploy for a given branch.
 * @param {string} branch  GitHub branch to deploy
 * @returns {Promise<string>} Deploy URL
 */
async function triggerDeploy(branch) {
  const projectId = process.env.VERCEL_PROJECT_ID;
  const orgId = process.env.VERCEL_ORG_ID;

  if (!projectId || !orgId) {
    throw new Error('VERCEL_PROJECT_ID and VERCEL_ORG_ID must be set');
  }

  const repoId = process.env.GITHUB_REPO_ID
    ? parseInt(process.env.GITHUB_REPO_ID, 10)
    : undefined;

  const res = await axios.post(
    `${VERCEL_API}/v13/deployments`,
    {
      name: 'student-management-ai-wpx8',
      gitSource: {
        type: 'github',
        ref: branch,
        repoId,
      },
    },
    { headers: { ...headers(), 'x-vercel-team-id': orgId } },
  );

  const deployId = res.data.id;
  console.log(`  🚀 Deploy triggered: ${deployId}`);

  return waitForDeploy(deployId, orgId);
}

/**
 * Poll until the deployment is ready or fails.
 * @param {string} deployId
 * @param {string} teamId
 * @returns {Promise<string>} Live URL
 */
async function waitForDeploy(deployId, teamId) {
  const maxAttempts = 30;
  const delayMs = 10_000;

  for (let attempt = 0; attempt < maxAttempts; attempt++) {
    await new Promise((r) => setTimeout(r, delayMs));

    const res = await axios.get(`${VERCEL_API}/v13/deployments/${deployId}`, {
      headers: { ...headers(), 'x-vercel-team-id': teamId },
    });

    const { readyState, url } = res.data;
    console.log(`  ⏳ Deploy status: ${readyState} (attempt ${attempt + 1}/${maxAttempts})`);

    if (readyState === 'READY') {
      return `https://${url}`;
    }
    if (['ERROR', 'CANCELED'].includes(readyState)) {
      throw new Error(`Deploy failed with state: ${readyState}`);
    }
  }

  throw new Error('Deploy timed out after 5 minutes');
}

/**
 * Get the latest deployment URL for the project (used as fallback).
 * @returns {Promise<string|null>}
 */
async function getLatestDeployUrl() {
  try {
    const res = await axios.get(`${VERCEL_API}/v6/deployments`, {
      headers: headers(),
      params: {
        projectId: process.env.VERCEL_PROJECT_ID,
        teamId: process.env.VERCEL_ORG_ID,
        limit: 1,
        state: 'READY',
      },
    });
    const latest = res.data.deployments?.[0];
    return latest ? `https://${latest.url}` : null;
  } catch {
    return null;
  }
}

module.exports = { triggerDeploy, getLatestDeployUrl };

File: python-backend/tools/vercel.py
import os
import httpx


def _headers() -> dict:
    return {
        "Authorization": f"Bearer {os.environ['VERCEL_TOKEN']}",
        "Content-Type": "application/json",
    }


async def trigger_deploy(branch_name: str) -> str:
    project_id = os.environ["VERCEL_PROJECT_ID"]
    org_id = os.environ.get("VERCEL_ORG_ID", "")
    repo_id_raw = os.environ.get("GITHUB_REPO_ID", "0")
    try:
        repo_id = int(repo_id_raw)
    except ValueError:
        repo_id = 0

    payload = {
        "name": project_id,
        "gitSource": {
            "type": "github",
            "repoId": repo_id,
            "ref": branch_name,
        },
    }
    params = {"teamId": org_id} if org_id else {}

    async with httpx.AsyncClient(timeout=60) as client:
        res = await client.post(
            "https://api.vercel.com/v13/deployments",
            headers=_headers(),
            json=payload,
            params=params,
        )
        data = res.json()
        if res.status_code not in (200, 201):
            raise RuntimeError(f"Vercel deploy failed {res.status_code}: {data}")
        deploy_id = data.get("id", "")
        url = data.get("url", "")
        print(f"  🚀 Deploy triggered: {deploy_id}")
        return f"https://{url}" if url and not url.startswith("http") else url


async def get_latest_deploy_url() -> str | None:
    project_id = os.environ["VERCEL_PROJECT_ID"]
    org_id = os.environ.get("VERCEL_ORG_ID", "")
    params = {"projectId": project_id, "limit": 1}
    if org_id:
        params["teamId"] = org_id

    async with httpx.AsyncClient(timeout=30) as client:
        res = await client.get(
            "https://api.vercel.com/v6/deployments",
            headers=_headers(),
            params=params,
        )
        deployments = res.json().get("deployments", [])
        if deployments:
            url = deployments[0].get("url", "")
            return f"https://{url}" if url and not url.startswith("http") else url
    return None
